pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/IndexLoadingConfig.java

```diff
@@ -0,0 +1,195 @@
+import com.linkedin.pinot.core.startree.StarTreeFormatVersion;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/**
+ * Table level index loading config.
+ */
+public class IndexLoadingConfig {
+  private static final int DEFAULT_REALTIME_AVG_MULTI_VALUE_COUNT = 2;
+
+  private ReadMode _readMode = ReadMode.DEFAULT_MODE;
+  private List<String> _sortedColumns = Collections.emptyList();
+  private Set<String> _invertedIndexColumns = new HashSet<>();
+  private Set<String> _noDictionaryColumns = new HashSet<>();
+  private SegmentVersion _segmentVersion = SegmentVersion.DEFAULT_VERSION;
+  private StarTreeFormatVersion _starTreeVersion = StarTreeFormatVersion.DEFAULT_VERSION;
+  private boolean _enableDefaultColumns = true;
+  private ColumnMinMaxValueGeneratorMode _columnMinMaxValueGeneratorMode = ColumnMinMaxValueGeneratorMode.DEFAULT_MODE;
+  private int _realtimeAvgMultiValueCount = DEFAULT_REALTIME_AVG_MULTI_VALUE_COUNT;
+
+  public IndexLoadingConfig(@Nullable InstanceDataManagerConfig instanceDataManagerConfig,
+      @Nullable AbstractTableConfig tableConfig) {
+    // Extract config from instance config
+    if (instanceDataManagerConfig != null) {
+      ReadMode instanceReadMode = instanceDataManagerConfig.getReadMode();
+      if (instanceReadMode != null) {
+        _readMode = instanceReadMode;
+      }
+
+      String instanceSegmentVersion = instanceDataManagerConfig.getSegmentFormatVersion();
+      if (instanceSegmentVersion != null) {
+        _segmentVersion = SegmentVersion.valueOf(instanceSegmentVersion.toLowerCase());
+      }
+
+      _enableDefaultColumns = instanceDataManagerConfig.isEnableDefaultColumns();
+
+      String avgMultiValueCount = instanceDataManagerConfig.getAvgMultiValueCount();
+      if (avgMultiValueCount != null) {
+        _realtimeAvgMultiValueCount = Integer.valueOf(avgMultiValueCount);
+      }
+    }
+
+    // Extract config from table indexing config
+    if (tableConfig != null) {
+      IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+      String tableReadMode = indexingConfig.getLoadMode();
+      if (tableReadMode != null) {
+        _readMode = ReadMode.getEnum(tableReadMode);
+      }
+
+      List<String> sortedColumns = indexingConfig.getSortedColumn();
+      if (sortedColumns != null) {
+        _sortedColumns = sortedColumns;
+      }
+
+      List<String> invertedIndexColumns = indexingConfig.getInvertedIndexColumns();
+      if (invertedIndexColumns != null) {
+        _invertedIndexColumns.addAll(invertedIndexColumns);
+      }
+
+      List<String> noDictionaryColumns = indexingConfig.getNoDictionaryColumns();
+      if (noDictionaryColumns != null) {
+        _noDictionaryColumns.addAll(noDictionaryColumns);
+      }
+
+      String tableSegmentVersion = indexingConfig.getSegmentFormatVersion();
```

Comment:
Not sure if we are preserving the existing logic of overriding the table config with server config if and only of the table config version is lower.